keps/sig-storage/1790-recover-resize-failure/README.md

```diff
@@ -0,0 +1,362 @@
+the user is using less storage than he/she actually is. To solve this
+problem - we propose that although users are allowed to reduce size of their
+PVC (as long as requested size >= `pvc.Status.Capacity`), quota calculation
+will only use previously requested higher value. This is covered in more
+detail in [Implementation](#implementation).
+
+
+### Goals
+
+- Allow users to cancel previously issued volume expansion requests, assuming they are not yet successful or have failed.
+- Allow users to retry expansion request with smaller value than original requested size in `pvc.Spec.Resources`, assuming previous requests are not yet successful or have failed.
+
+### Non-Goals
+
+- As part of this KEP we do not intend to actually support shrinking of underlying volume.
+
+## Proposal
+
+As part of this proposal, we are mainly proposing three changes:
+
+1. Relax API validation on PVC update so as reducing `pvc.Spec.Resoures` is allowed, as long as requested size is `>=pvc.Status.Capacity`.
+2. Add a new field in PVC API called - `pvc.Spec.AllocatedResources` for the purpose of tracking used storage size. This field is only allowed to increase.
+3. Update quota code to use `max(pvc.Spec.Resources, pvc.Spec.AllocatedResources)` when evaluating usage for PVC.
+
+### Implementation
+
+We propose that by relaxing validation on PVC update to allow users to reduce `pvc.Spec.Resources`, it becomes possible to cancel previously issued expansion requests or retry expansion with a lower value if previous request has not been successful. In general - we know that volume plugins are designed to never perform actual shrinking of the volume, for both in-tree and CSI volumes. Moreover if a previously issued expansion has been successful and user
+reduces the PVC request size, for both CSI and in-tree plugins they are designed to return a successful response with NO-OP. So, reducing requested size will be a safe operation and will never result in data loss or actual shrinking of volume.
+
+We however do have a problem with quota calculation because if a previously issued expansion is successful but is not recorded(or partially recorded) in api-server and user reduces requested size of the PVC, then quota controller will assume it as actual shrinking of volume and reduce used storage size by the user(incorrectly). Since we know actual size of the volume only after performing expansion(either on node or controller), allowing quota to be reduced on PVC size reduction will allow an user to abuse the quota system.
+
+To solve aforementioned problem - we propose that, a new field will be added to PVC, called `pvc.Spec.AllocatedResources`. This field is only allowed to increase and will be set by the api-server to value in `pvc.Spec.Resources` as long as `pvc.Spec.Resources > pvc.Spec.AllocatedResources`.  Quota controller code will be updated to use `max(pvc.Spec.Resources, pvc.Spec.AllocatedResources)` when calculating usage of the PVC under questions.
```

Comment:
1. Can you explain why `pvc.Status.Capacity` can't be used for quota allocation? Why doesn't that always reflect the actual capacity?
2. Why is `AllocatedResources` under `pvc.Spec` and not `pvc.Status`?
3. Is `AllocatedResources` the best name? As I understand it, it will contain the max size requested (regardless of actual size), right? "Allocated" suggests that is the real size allocated on the backend.

Reply:
> Can you explain why pvc.Status.Capacity can't be used for quota allocation? Why doesn't that always reflect the actual capacity?

`pvc.Status.Capacity` can't be used for tracking quota because `pvc.Status.Capacity` is calculated after binding happens, which could be when pod is started.  This would allow an user to overcommit because quota won't reflect accurate value until PVC is bound to a PV.

> Why is AllocatedResources under pvc.Spec and not pvc.Status?

There are two reasons:
1. `AllocatedResources` is not volume size but more like whatever user has requested and towards which resize-controller was working to reconcile.  It is possible that user has requested smaller size since then but that does not changes the fact that resize-controller has already tried to expand to `AllocatedResources` and might have partially succeeded. So `AllocatedResources` is maximum user requested size for this volume and does not reflect actual volume size of the PV.

2. Following from https://github.com/kubernetes/enhancements/pull/1342/files (pointed by Jordan), it also uses `spec.containers[i].ResourcesAllocated` for tracking user requests, so keeping this in PVC makes it somewhat consistent.

> Is AllocatedResources the best name? As I understand it, it will contain the max size requested (regardless of actual size), right? "Allocated" suggests that is the real size allocated on the backend.

There are two reasons. First is to follow convention of vertical pod scaling KEP and then second one is - `AllocatedResources` does mean that this is the size resize controller (or pv) has tried to reconcile towards.  It is possible that user reduced requested size and then resize controller stopped reconciliation but it may already have succeeded, so `AllocatedResources` does sound like okay name.

Reply:
This is good context. Can you add it to the doc? Otherwise LGTM

Reply:
added to the KEP. One under alternatives considered and another one about naming.